### PR TITLE
Update loss free & loss control load balance

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Please see this [screenshot](https://screenshot.googleplex.com/8N88e9LAmjAZLuo) for high-level pseudo-code, b/461812380.

* Add `routed_bias_update_rate` config to indicate the speed of bias term updates
* Add helper function `calculate_load_balance_updates()` and pass it via `sow` to param updates.  
* Add helper function `update_state_param()` to update or replace the param values.
* Enable **optional** loss control load balance mechanism in other MoE models, like DeepSeek, GPT-OSS, and Llama4 family. We have enabled them in Mixtral & Qwen3 family already.

# Tests
* Expect runner tests are green to catch any compile issue
* Add unit tests for `calculate_load_balance_updates()` and `update_state_param()` functions
* Functional tests:
  * Only DeepSeek models support loss free case, expected - [test](https://paste.googleplex.com/4603411184091136)
  * Optionally support loss control case for MoE models - [test](https://paste.googleplex.com/5112262668976128)
* Performance check before vs. after to run DeepSeek v2 on synthetic dataset `per_device_batch_size=4, max_target_length=4096`

```
# before the change on Main
I1228 23:27:36.115740 140683853757568 metric_logger.py:180] completed step: 14, seconds: 2.557, TFLOP/s/device: 105.114, Tokens/s/device: 6406.837, total_weights: 65536, loss: 10.739

# after the change with this commit 
I1228 23:23:18.459741 140009441311872 metric_logger.py:180] completed step: 14, seconds: 2.557, TFLOP/s/device: 105.124, Tokens/s/device: 6407.436, total_weights: 65536, loss: 10.739
```

* Tested locally to verify the bias update as expectation - [full logs](https://paste.googleplex.com/5411290761789440#l=1974)

Two expected behaviors:
1) before_the_change + bias_updates = after_the_change (expect slight delta for very small values). For instance, -0.000450134 - 0.0005  = -0.000950134 ~= -0.00094986 with reasonable tolerance (|-0.000950134+0.00094986| = 0.000000274).
2) after_the_change in previous layer = bias in next layer

```
before the change: [[-0.000450134]
 [-0.000457764]
 [-0.000537872]
 [0.000463486]
 [0.000480652]
 ...

after the change: [[-0.00094986]
 [-0.000957489]
 [-0.0010376]
 [0.000965118]
 [0.000984192]
...

bias_updates: [-0.0005 -0.0005 -0.0005  0.0005  0.0005 ...]

bias: [-0.00094986 -0.000957489 -0.0010376 0.000965118 0.000984192...
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: docs/reference/core_concepts/moe_configuration.md [hunk 1]
```diff
@@ -59,6 +59,8 @@ Dropping:
 
 `routed_bias`: If enabled, adds a learnable bias term to the gate logits to facilitate load balancing.
 
+`routed_bias_update_rate`: Defines the update rate to routed bias term above.
```
Comment (NuojCheng) [inline docs/reference/core_concepts/moe_configuration.md]: maybe specify this flag only works for DS family?

File: src/MaxText/configs/types.py [hunk 2]
```diff
@@ -552,7 +552,7 @@ class MoEGeneral(BaseModel):
   num_experts: PositiveInt = Field(1, description="The total number of experts in each MoE layer.")
   num_experts_per_tok: PositiveInt = Field(1, description="The number of experts to route each token to.")
   capacity_factor: float = Field(-1.0, description="Expert capacity factor. If < 0, no token dropping.")
-  load_balance_loss_weight: NonNegativeFloat = Field(0.01, description="Weight for the load balancing auxiliary loss.")
+  load_balance_loss_weight: NonNegativeFloat = Field(0.0, description="Weight for the load balancing auxiliary loss.")
```
Comment (NuojCheng) [inline src/MaxText/configs/types.py:555]: Why we need to adjust load_balance_loss_weight default value?

**Review by NuojCheng — APPROVED** [review]
Thank you Ran! LGTM

File: docs/reference/core_concepts/moe_configuration.md (hunk 1 above)
Comment (RissyRan) [reply docs/reference/core_concepts/moe_configuration.md]: Good point, I will add that in the description.

File: src/MaxText/configs/types.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/configs/types.py:555]: Yeah, I noticed that we should set it as 0.0, which is disabled by default. Let ask customers to adjust the value they'd like to have.

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -714,6 +714,38 @@ def get_nested_value(dictionary, nested_key, default=None):
   return current_level
 
 
+def update_state_param(state, target_path, value, mode="add"):
```
Comment (gobbleturk) [inline src/MaxText/maxtext_utils.py]: This is currently a very general function which only has one use case. Have you considered narrowing it in scope to only support the one use case now - see https://engdoc.corp.google.com/eng/doc/tott/episodes/737.md?cl=head

**Review by gobbleturk — APPROVED** [review]

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/maxtext_utils.py]: Np! I will narrow it down, and remove the tests accordingly.

Comment (RissyRan) [reply src/MaxText/maxtext_utils.py]: Updated! Thanks for catching it!
